File: packages/mirascript/src/vm/lib/global/math/const.ts
import { DiagnosticCode } from '@mirascript/constants';
import { VmLib } from '../../helpers.js';

export const PI = VmLib(Math.PI, {
    summary: '圆周率',
    returnsType: 'number',
});
export const E = VmLib(Math.E, {
    summary: '自然对数的底数',
    returnsType: 'number',
});
export const pi = VmLib(Math.PI, {
    summary: '圆周率',
    returnsType: 'number',
    deprecated: {
        use: 'PI',
        message: DiagnosticCode.PreferUppercaseConstant,
    },
});
export const e = VmLib(Math.E, {
    summary: '自然对数的底数',
    returnsType: 'number',
    deprecated: {
        use: 'E',
        message: DiagnosticCode.PreferUppercaseConstant,
    },
});
export const SQRT1_2 = VmLib(Math.SQRT1_2, {
    summary: '½ 的平方根',
    returnsType: 'number',
});
export const SQRT2 = VmLib(Math.SQRT2, {
    summary: '2 的平方根',
    returnsType: 'number',
});
export const LN2 = VmLib(Math.LN2, {
    summary: '2 的自然对数',
    returnsType: 'number',
});
export const LN10 = VmLib(Math.LN10, {
    summary: '10 的自然对数',
    returnsType: 'number',
});
export const LOG2E = VmLib(Math.LOG2E, {
    summary: 'e 以 2 为底的对数',
    returnsType: 'number',
});
export const LOG10E = VmLib(Math.LOG10E, {
    summary: 'e 以 10 为底的对数',
    returnsType: 'number',
});

File: crates/python/mirascript/_vm/lib/vm_global/math/const.py
from __future__ import annotations
from typing_extensions import Final
import math

__all__ = [
    "pi",
    "PI",
    "e",
    "E",
    "SQRT1_2",
    "SQRT2",
    "LN2",
    "LN10",
    "LOG2E",
    "LOG10E",
]

pi: Final = math.pi
PI: Final = pi
e: Final = math.e
E: Final = e

SQRT1_2: Final = 2**-0.5
SQRT2: Final = 2**0.5
LN2: Final = 0.6931471805599453  # math.log(2)
LN10: Final = 2.302585092994046  # math.log(10)
LOG2E: Final = 1.4426950408889634  # math.log(e, 2)
LOG10E: Final = 0.4342944819032518  # math.log(e, 10)
